articles/qos_configurability.md

```diff
@@ -0,0 +1,399 @@
+Some concerns were raised by the community:
+1. It adds a different mechanism to the ones that already exist.
+    Newcomers have to learn yet another thing.
+1. QoS settings are not conceptually different to other parameters, e.g.: publishing rate.
+1. It adds a new mechanism to configure QoS profiles.
+
+Based on that feedback, this PR will explore a design based on parameters.
+
+## Introduction
+
+Current code will not allow QoS reconfigurability if not changed:
+
+```cpp
+node->create_publisher(
+  "chatter",
+  KeepLast{10});
+```
+
+To make reconfigurability easy, only a flag is needed to automatically create the parameters:
+
+```cpp
+node->create_publisher(
+  "chatter",
+  KeepLast{10},
+  QosOverridingOptions{true});  // allow_reconfigurable all qos
+```
+
+That will automatically declare the parameters for reconfiguring the QoS policies that can be overridden in the following way:
+
+```yaml
+/my/ns/node_name:
+  ros__parameters:
+    qos_profiles:  # group for all the qos settings
+      publisher:
+        'my/fully/qualified/topic/name/here': 
+          reliability: reliable
+          history_depth: 100
+          history: keep_last
+```
+
+A callback to accept/reject the configured profile can also be specified:
+
+```cpp
+node->create_publisher(
+  "chatter",
+  KeepLast(10),
+  [] (const QoSProfile & qos) -> bool {
+      return qos.is_reliable();  // only accept reliable communication, we should add getters to rclcpp::QoSProfile class for this.
+  });
+```
+
+The user provided QoS callback will be internally used as a [parameters callback](https://github.com/ros2/rclcpp/blob/3a4ac0ca2093d12035070443692798b0c9f9da3a/rclcpp/include/rclcpp/node_interfaces/node_parameters_interface.hpp#L183), but we provide a more user-friendly interface.
+
+## Analysis of parameter API features and limitations
+
+### Read-only parameters
+
+ROS 2 currently provides read only parameters.
+They can be modified when constructing a node by providing overrides (e.g. `--ros-args -p <param_name> <param_value>`), but they cannot be dynamically changed after the node was constructed.
+
+This perfectly matches the goal of the proposal of not making QoS settings reconfigurable during runtime.
+
+Note: parameter overrides can be constructed dynamically and then used to create a node, but it's not possible to modify the parameter after the node construction.
+This allows, for example, to pass parameter overrides when dynamically adding a node to a component container.
+
+### Parameter events
+
+Up to ROS 2 Foxy, read-only parameters generate a parameter event with its initial value when declared.
+If we start declaring parameters for each QoS policies of many entities, the amount of parameter events at startup time will significantly increase, as the declaration of each parameter generates an event.
+
+To avoid this issue, a way to opt-out parameter events when declaring a parameter could be added.
```

Comment:
I think that's reasonable.

Do you mean?
- Turn off parameter events for read-only parameters.
- Turn off by default parameter events generated when declaring a parameter, but continue generating events when a parameter is set.
- Turn off by default all parameter events.

`rclcpp` currently depends on parameters events to make `use_sim_time` works, but that's fixable.

Reply:
I mean to turn off parameter events for any parameters used internally by the ROS infrastructure, in the way these QoS ones are. I don't know how many of those there are at this time, though.

My reasoning is that a node author is going to care about events for parameters they declare themselves, but will find any additional events that they have to handle in some way to be noise.

Reply:
Ah, that makes sense.
I miss-read your previous comment.

Reply:
I actually think we shouldn't do this. I think we should treat all of the parameters the same. It would be confusing to me as a user to see that some parameters cause events and others do not.

Answering the question "which parameters generate events" is not very trivial then.

That being said, if we do this, then it should be possible for the user to have parameters that do not generate events as well, i.e. we should dog food this feature.